File: decorators.py
import time


def generic_decorator(func):
    def modifier(*args):
        return func(*args)
    return modifier


def timer(func):
    """
        \*args allow us to pass any number of anonymous arguments,
        preserving the function's original implementation.
    """
    def wrapper(*args):
        start_time = time.process_time()
        result = func(*args)
        elapsed = time.process_time() - start_time
        print(f"Python: {elapsed:.3f} ")
        return result
    return wrapper


def memoize(func):
    cache = {}

    def wrapper(*args):
        if args in cache:
            print(f"Getting result from cache for {args}")
            return cache[args]
        result = func(*args)
        cache[args] = result
        return result
    return wrapper


@timer
@memoize
def cpu_spinner(x):
    values = [x for _ in range(x)]
    return sum(values)


cpu_spinner(100000000)
cpu_spinner(100000000)

"""
Adding authentication to functions:

We can create a decorator function that checks if the user is authenticated before allowing them to execute a function. This can be useful in creating secure APIs or web applications.

python
Copy code
def authenticate_decorator(func):
    def wrapper(*args, **kwargs):
        if is_authenticated():
            return func(*args, **kwargs)
        else:
            raise Exception("User is not authenticated!")
    return wrapper

@authenticate_decorator
def secure_function():
    return "Sensitive information"

def is_authenticated():
    return False

secure_function()
"""

File: decorators.js
function timer(func) {
  function wrapper(...args) {
    console.time("JavaScript");
    const result = func(...args);
    console.timeEnd("JavaScript");
    return result;
  }
  return wrapper;
}

function memoize(func) {
  // initialize mutable structure
  let cache = {};

  function wrapper(...args) {
    const key = args.toString();
    // return early with result if it's stored
    if (key in cache) return cache[key];
    const result = func(...args);
    cache[key] = result;
    return result;
  }
  return wrapper;
}

function cpuSpinner(x) {
  return [...Array.from({ length: x }, (x) => x)].reduce(
    (sum, val) => (sum += val),
    0
  );
}

function cpuSpinnerIter(x) {
  const nums = [...Array.from({ length: x }, (x) => x)];
  let total = 0;
  for (let i = 0; i < nums.length; i++) {
    total += nums[i];
  }
}

const timedMemo = timer(memoize(cpuSpinner));

timedMemo(100000000);
timedMemo(100000000);

const timedIterMemo = timer(memoize(cpuSpinnerIter));

timedIterMemo(100000000);
timedIterMemo(100000000);
